### PR TITLE
[Main] Automate database corruption fix caused by out of sync txdb.

### PR DESCRIPTION
**Pull request by presstab**
**Summary:**
- Refactor the automated db corruption repair code to be based on comparing `pcoinsTip`'s best block, with the best block loaded from block index db. If the wallet client is closed forcefully before it completes `CheckBlock()` it causes a state where the block index is stored to disk but the transaction data is not. The automated repair code will detect on initialization when there is a mismatch between the two databases. If the transaction database is behind the block index database, each block from the best height in the tx database up until the best block of the block index database will be parsed and all transactions recorded and flushed to disk.

- On block 908000 a block hash checkpoint is added. If the validation process fails to validate block 908000 because it was previously marked invalid, the user will not need to use the `reconsiderblock` RPC command to clear the status. The code will now see that they are on block 908000, check if the blockhash matches the checkpoint, and reprocess the block.

- On initial block download (IDB), the wallet client is not triggered to calculate the invalid outpoint map. The invalid outpoint map is used to calculate the accumulator checkpoint for block 908000, so if it never gets populated, then a checkpoint will be generated that does not match the blockheader's accumulator checkpoint, which will result in the block being rejected. Before calculating the checkpoint for 908000 the code now forces the population of the invalid outpoint map, avoiding this scenario.

- Around 908000 many nodes have a `zerocoinDB` that is not in its proper state with accumulators properly calculated. If an accumulator checkpoint is rejected in close proximity to block 908000, then a reindexing of the accumulators will automatically be done, and the checkpoint recalculated in an attempt to fix this issue.

- Deprecate `spork_11` and replace it with a value called by `Params()`

**Tested Scenarios:**
- Sync from scratch - works.
- Reindex - works.
- Blockchain snapshot that synced from scratch, but got stuck at block 907999 - fixed by running the new wallet.
- Blockchain snapshot that was experiencing `assert(hashPrevBlock == view.GetBestBlock());` from `ConnectBlock()` due to having a txdb that was *behind* the blockchain db : fixed by running the new wallet. Note: This is the issue that the majority of nodes that are stuck on block 907999 will have because the wallet froze in the middle of `ConnectBlock()` causing the txdb to not be flushed to disk.
- Blockchain snapshot that was experiencing `assert(hashPrevBlock == view.GetBestBlock());` from `ConnectBlock()` due to having a txdb that was *ahead* of the blockchain db : fixed by restarting wallet with `-reindex`.

### CODE REVIEW

**Comment by presstab** [comment]
@Mrs-X good find. Changed to:
` if (pindex && Checkpoints::CheckBlock(pindex->nHeight - 1, block.hashPrevBlock, true)) {`

**Comment by Fuzzbawls** [comment]
confirmed that a fresh sync using a bootstrap file now properly handles the accumulator/supply recalculations at block `908000`

File: src/main.cpp [hunk 1]
```diff
@@ -3196,6 +3198,59 @@ bool RecalculatePIVSupply(int nHeightStart)
     return true;
 }
 
+bool ReindexAccumulators(list<uint256> listMissingCheckpoints, string& strError)
```
[the remark this answers is not in the document]
Comment (presstab) [reply src/main.cpp]: The way I have it coded, it erases a checkpoint from `listMissingCheckpoints` once it finds/processes it. Unable to do that with const ref.

File: src/main.cpp [hunk 2]
```diff
@@ -3196,6 +3198,59 @@ bool RecalculatePIVSupply(int nHeightStart)
...
+{
+    // PIVX: recalculate Accumulator Checkpoints that failed to database properly
+    if (!listMissingCheckpoints.empty() && chainActive.Height() >= Params().Zerocoin_StartHeight()) {
```
[the remark this answers is not in the document]
Comment (presstab) [reply src/main.cpp]: That is exactly what this code is. You were commenting on the removed code above.

It was formerly not a function and just a one time procedure ran in init.cpp. Now it is accessible anytime with `ReindexAccumulators()`

File: src/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonspock) [reply src/main.cpp]: OK, then why not just regular reference?

File: src/main.cpp [hunk 3]
```diff
@@ -3453,17 +3508,42 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+                    if (!count(listCheckpoints.begin(), listCheckpoints.end(), pindexCheckpoint->nAccumulatorCheckpoint))
+                        listCheckpoints.emplace_back(pindexCheckpoint->nAccumulatorCheckpoint);
+
+                    pindexCheckpoint = chainActive.Next(pindexCheckpoint);
+                    if (!pindexCheckpoint)
+                        break;
+                }
+
+                string strError;
+                if (!ReindexAccumulators(listCheckpoints, strError) || !CalculateAccumulatorCheckpoint(pindex->nHeight, nCheckpointCalculated))
```
[the remark this answers is not in the document]
Comment (presstab) [reply src/main.cpp]: Empty list will skip the procedure within the function and return true.

File: src/main.cpp [hunk 4]
```diff
@@ -3453,17 +3508,42 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+                    return state.DoS(100, error("ConnectBlock() : failed to recalculate accumulator checkpoint"));
+            } else {
+                return state.DoS(100, error("ConnectBlock() : failed to calculate accumulator checkpoint"));
+            }
+        }
 
         if (nCheckpointCalculated != block.nAccumulatorCheckpoint) {
             LogPrintf("%s: block=%d calculated: %s\n block: %s\n", __func__, pindex->nHeight, nCheckpointCalculated.GetHex(), block.nAccumulatorCheckpoint.GetHex());
             return state.DoS(100, error("ConnectBlock() : accumulator does not match calculated value"));
```
[the remark this answers is not in the document]
Comment (presstab) [reply src/main.cpp]: Good suggestion. Adding an error message that will feed into the reindex popup.

pull_405_popup

File: src/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (presstab) [reply src/main.cpp]: Will do

**Comment by Mrs-X** [comment]
utACK and merging...

**Pull request merged**